delegation/src/Pos/Binary/Delegation.hs

```diff
@@ -6,19 +6,34 @@ module Pos.Binary.Delegation
 
 import           Universum
 
-import           Pos.Binary.Class (Bi (..), Cons (..), Field (..), deriveSimpleBiCxt)
+import qualified Data.Set as Set
+
+import           Pos.Binary.Class (Bi (..), encodeListLen, enforceSize)
 import           Pos.Binary.Core ()
 import           Pos.Binary.Crypto ()
 import           Pos.Communication.Types.Relay (DataMsg (..))
-import           Pos.Core (ProxySKHeavy, StakeholderId)
+import           Pos.Core (ProxySKHeavy)
 import           Pos.Core.Configuration (HasConfiguration)
+import           Pos.Crypto (PublicKey, pskIssuerPk)
 import           Pos.Delegation.Types (DlgUndo (..))
 
-deriveSimpleBiCxt [t|HasConfiguration|] ''DlgUndo [
-    Cons 'DlgUndo [
-        Field [| duPsks            :: [ProxySKHeavy]        |],
-        Field [| duPrevEpochPosted :: HashSet StakeholderId |]
-    ]]
+instance HasConfiguration => Bi DlgUndo where
+    encode DlgUndo{..} =
+        encodeListLen 2 <>
+        encode duPsks <>
+        encode duPrevEpochPosted
+    decode = do
+        enforceSize "DlgUndo" 2
+
+        (duPsks :: [ProxySKHeavy]) <- decode
+        let duPsksSet :: Set PublicKey
+            duPsksSet = Set.fromList $ map pskIssuerPk duPsks
+        when (length duPsks /= Set.size duPsksSet) $
+              fail "DlgUndo.duPsks is not a set: it has duplicates in issuers"
```

Comment:
AFAIU, changes in this file ought to be reverted.

Reply:
You're right, but I want to leave this one.